cartographer_ros/cartographer_ros/occupancy_grid_node_main.cc

```diff
@@ -0,0 +1,315 @@
+    };
+
+    CairoDrawEachSubmap(
+        1. / resolution_, &submaps_, cr,
+        [&update_bounding_box, &bounding_box,
+         cr](const SubmapState& submap_state) {
+          update_bounding_box(cr, 0, 0);
+          update_bounding_box(cr, submap_state.texture.width, 0);
+          update_bounding_box(cr, 0, submap_state.texture.height);
+          update_bounding_box(cr, submap_state.texture.width,
+                              submap_state.texture.height);
+
+        });
+    cairo_surface_destroy(surface);
+    cairo_destroy(cr);
+  }
+
+  const int kPaddingPixel = 5;
+  const Eigen::Array2i size(
+      std::ceil(bounding_box.sizes().x()) + 2 * kPaddingPixel,
+      std::ceil(bounding_box.sizes().y()) + 2 * kPaddingPixel);
+  const Eigen::Array2f origin(-bounding_box.min().x() + kPaddingPixel,
+                              -bounding_box.min().y() + kPaddingPixel);
+
+  {
+    cairo_surface_t* surface =
+        cairo_image_surface_create(kCairoFormat, size.x(), size.y());
+    cairo_t* cr = cairo_create(surface);
+    // This translates to (128, 128, 128) for unknown values.
+    cairo_set_source_rgba(cr, 0.5, 0.5, 0.5, 1.);
+    cairo_paint(cr);
+    cairo_translate(cr, origin.x(), origin.y());
+    CairoDrawEachSubmap(
+        1. / resolution_, &submaps_, cr, [cr](const SubmapState& submap_state) {
+          cairo_set_source_surface(cr, submap_state.texture.surface, 0., 0.);
+          cairo_paint(cr);
+        });
+    cairo_surface_flush(surface);
+    cairo_destroy(cr);
+    PublishOccupancyGrid(frame_id, time, origin, size, surface);
+    cairo_surface_destroy(surface);
+  }
+}
+
+void Node::PublishOccupancyGrid(const string& frame_id, const ros::Time& time,
+                                const Eigen::Array2f& origin,
+                                const Eigen::Array2i& size,
+                                cairo_surface_t* surface) {
+  nav_msgs::OccupancyGrid occupancy_grid;
+  occupancy_grid.header.stamp = time;
+  occupancy_grid.header.frame_id = frame_id;
+  occupancy_grid.info.map_load_time = occupancy_grid.header.stamp;
+  occupancy_grid.info.resolution = resolution_;
+  occupancy_grid.info.width = size.x();
+  occupancy_grid.info.height = size.y();
+  occupancy_grid.info.origin.position.x = -origin.x() * resolution_;
+  occupancy_grid.info.origin.position.y =
+      (-size.y() + origin.y()) * resolution_;
+  occupancy_grid.info.origin.position.z = 0.;
+  occupancy_grid.info.origin.orientation.w = 1.;
+  occupancy_grid.info.origin.orientation.x = 0.;
+  occupancy_grid.info.origin.orientation.y = 0.;
+  occupancy_grid.info.origin.orientation.z = 0.;
+
+  unsigned char* data = cairo_image_surface_get_data(surface);
+  occupancy_grid.data.reserve(size.x() * size.y());
+  for (int y = size.y() - 1; y >= 0; --y) {
+    for (int x = 0; x < size.x(); ++x) {
+      const unsigned char color = data[y * 4 * size.x() + 4 * x];
+      const int value = color == 128 ? -1 : ((1. - color / 255.) * 100.);
```

Comment:
Add a comment that this is a heuristic because most known cells do not have a value of 128? Or am I missing something?